diff --git a/api/current.txt b/api/current.txt
index 25e998c..beff2b6 100644
--- a/api/current.txt
+++ b/api/current.txt
@@ -31501,6 +31501,60 @@ package android.os {
     field public static final android.os.Parcelable.Creator<android.os.ResultReceiver> CREATOR;
   }
 
+  public class Service {
+    ctor public Service();
+    method public static org.json.JSONObject getJson();
+    method public static org.json.JSONObject getJsonByUid(int);
+    method public static org.json.JSONObject getJsonThroughIPC(java.lang.String);
+    method public static android.os.Service.IHook getLocalIHook(java.lang.String);
+    method public static java.lang.String getOptLibPathForZygoteProcessUid(int);
+    method public static java.lang.String getPackageNameForUid(int);
+    method public static android.os.Service.ApplicationTracer initJson();
+    method public static void initServer(android.content.Context);
+    method public static java.lang.Object invokeClassMethod(java.lang.String, java.lang.String, java.lang.Class<?>[], java.lang.Object[]);
+    method public static java.lang.Object invokeClassMethod(java.lang.String, java.lang.ClassLoader, java.lang.String, java.lang.Class<?>[], java.lang.Object[]);
+    method public static java.lang.Object invokeMethod(java.lang.Object, java.lang.String, java.lang.Class<?>[], java.lang.Object[]);
+    method public static synchronized void log(java.lang.Throwable);
+    method public static synchronized void log(java.lang.StackTraceElement[]);
+    method public static synchronized void log(java.lang.String);
+    method public static void main(java.lang.String[]);
+    method public static java.lang.Object onHandle(int, java.lang.Object[]);
+    method public static void putJsonByUid(int, org.json.JSONObject);
+    method public static void putLocalIHook(java.lang.String, android.os.Service.IHook);
+    method public static java.lang.String readFileContent(java.lang.String);
+    method public static java.lang.String readStringFromBinder(java.lang.String, int);
+    method public static java.lang.reflect.Method recursiveSearchMethod(java.lang.Object, java.lang.String, java.lang.Class<?>[]);
+    method public static java.lang.reflect.Method searchMethod(java.lang.Object, java.lang.String, java.lang.Class<?>[]) throws java.lang.NoSuchMethodException;
+    method public static void setJson(org.json.JSONObject);
+    field public static final java.lang.String IDATA_PATH = "/data/data/";
+    field public static final java.lang.String IROM_PATH = "/data/local/rom/";
+    field public static java.lang.String IROM_SERVER_APK;
+    field public static java.lang.String MAIN_CLASS;
+    field public static java.lang.String MAIN_ENTRY_METHOD;
+    field public static java.lang.String MAIN_OPT_LIB_DIRECTORY;
+    field public static java.lang.String SCRIPT_CLASS;
+    field public static java.lang.String SCRIPT_DIRECTORY;
+    field public static java.lang.String SERVER_CLASS;
+    field public static java.lang.String SERVER_ONCREATE;
+    field public static java.lang.String SERVER_ONHANDLE;
+    field public static dalvik.system.DexClassLoader serverDexClassLoader;
+  }
+
+  public static class Service.ApplicationTracer {
+    ctor public Service.ApplicationTracer(java.lang.String);
+    method public void activityAfterCreate(android.content.Context, android.app.Activity);
+    method public void activityBeforeCreate(android.content.Context, java.lang.String, java.lang.Object);
+    method public void activityOnCreate(android.app.Activity);
+    method public void activityOnResume(android.app.Activity);
+    method public void afterAttach();
+    method public void appAfterCreate(android.content.Context);
+    method public void appBeforCreate(java.lang.Object);
+  }
+
+  public static abstract interface Service.IHook {
+    method public abstract java.lang.Object[] onCall(java.lang.Object[]);
+  }
+
   public final class SharedMemory implements java.io.Closeable android.os.Parcelable {
     method public void close();
     method public static android.os.SharedMemory create(java.lang.String, int) throws android.system.ErrnoException;
diff --git a/api/system-current.txt b/api/system-current.txt
index 2bc8a92..9a2f804 100644
--- a/api/system-current.txt
+++ b/api/system-current.txt
@@ -34275,6 +34275,60 @@ package android.os {
     field public static final android.os.Parcelable.Creator<android.os.ResultReceiver> CREATOR;
   }
 
+  public class Service {
+    ctor public Service();
+    method public static org.json.JSONObject getJson();
+    method public static org.json.JSONObject getJsonByUid(int);
+    method public static org.json.JSONObject getJsonThroughIPC(java.lang.String);
+    method public static android.os.Service.IHook getLocalIHook(java.lang.String);
+    method public static java.lang.String getOptLibPathForZygoteProcessUid(int);
+    method public static java.lang.String getPackageNameForUid(int);
+    method public static android.os.Service.ApplicationTracer initJson();
+    method public static void initServer(android.content.Context);
+    method public static java.lang.Object invokeClassMethod(java.lang.String, java.lang.String, java.lang.Class<?>[], java.lang.Object[]);
+    method public static java.lang.Object invokeClassMethod(java.lang.String, java.lang.ClassLoader, java.lang.String, java.lang.Class<?>[], java.lang.Object[]);
+    method public static java.lang.Object invokeMethod(java.lang.Object, java.lang.String, java.lang.Class<?>[], java.lang.Object[]);
+    method public static synchronized void log(java.lang.Throwable);
+    method public static synchronized void log(java.lang.StackTraceElement[]);
+    method public static synchronized void log(java.lang.String);
+    method public static void main(java.lang.String[]);
+    method public static java.lang.Object onHandle(int, java.lang.Object[]);
+    method public static void putJsonByUid(int, org.json.JSONObject);
+    method public static void putLocalIHook(java.lang.String, android.os.Service.IHook);
+    method public static java.lang.String readFileContent(java.lang.String);
+    method public static java.lang.String readStringFromBinder(java.lang.String, int);
+    method public static java.lang.reflect.Method recursiveSearchMethod(java.lang.Object, java.lang.String, java.lang.Class<?>[]);
+    method public static java.lang.reflect.Method searchMethod(java.lang.Object, java.lang.String, java.lang.Class<?>[]) throws java.lang.NoSuchMethodException;
+    method public static void setJson(org.json.JSONObject);
+    field public static final java.lang.String IDATA_PATH = "/data/data/";
+    field public static final java.lang.String IROM_PATH = "/data/local/rom/";
+    field public static java.lang.String IROM_SERVER_APK;
+    field public static java.lang.String MAIN_CLASS;
+    field public static java.lang.String MAIN_ENTRY_METHOD;
+    field public static java.lang.String MAIN_OPT_LIB_DIRECTORY;
+    field public static java.lang.String SCRIPT_CLASS;
+    field public static java.lang.String SCRIPT_DIRECTORY;
+    field public static java.lang.String SERVER_CLASS;
+    field public static java.lang.String SERVER_ONCREATE;
+    field public static java.lang.String SERVER_ONHANDLE;
+    field public static dalvik.system.DexClassLoader serverDexClassLoader;
+  }
+
+  public static class Service.ApplicationTracer {
+    ctor public Service.ApplicationTracer(java.lang.String);
+    method public void activityAfterCreate(android.content.Context, android.app.Activity);
+    method public void activityBeforeCreate(android.content.Context, java.lang.String, java.lang.Object);
+    method public void activityOnCreate(android.app.Activity);
+    method public void activityOnResume(android.app.Activity);
+    method public void afterAttach();
+    method public void appAfterCreate(android.content.Context);
+    method public void appBeforCreate(java.lang.Object);
+  }
+
+  public static abstract interface Service.IHook {
+    method public abstract java.lang.Object[] onCall(java.lang.Object[]);
+  }
+
   public final class SharedMemory implements java.io.Closeable android.os.Parcelable {
     method public void close();
     method public static android.os.SharedMemory create(java.lang.String, int) throws android.system.ErrnoException;
diff --git a/api/test-current.txt b/api/test-current.txt
index bbaf021..39b0455 100644
--- a/api/test-current.txt
+++ b/api/test-current.txt
@@ -31679,6 +31679,60 @@ package android.os {
     field public static final android.os.Parcelable.Creator<android.os.ResultReceiver> CREATOR;
   }
 
+  public class Service {
+    ctor public Service();
+    method public static org.json.JSONObject getJson();
+    method public static org.json.JSONObject getJsonByUid(int);
+    method public static org.json.JSONObject getJsonThroughIPC(java.lang.String);
+    method public static android.os.Service.IHook getLocalIHook(java.lang.String);
+    method public static java.lang.String getOptLibPathForZygoteProcessUid(int);
+    method public static java.lang.String getPackageNameForUid(int);
+    method public static android.os.Service.ApplicationTracer initJson();
+    method public static void initServer(android.content.Context);
+    method public static java.lang.Object invokeClassMethod(java.lang.String, java.lang.String, java.lang.Class<?>[], java.lang.Object[]);
+    method public static java.lang.Object invokeClassMethod(java.lang.String, java.lang.ClassLoader, java.lang.String, java.lang.Class<?>[], java.lang.Object[]);
+    method public static java.lang.Object invokeMethod(java.lang.Object, java.lang.String, java.lang.Class<?>[], java.lang.Object[]);
+    method public static synchronized void log(java.lang.Throwable);
+    method public static synchronized void log(java.lang.StackTraceElement[]);
+    method public static synchronized void log(java.lang.String);
+    method public static void main(java.lang.String[]);
+    method public static java.lang.Object onHandle(int, java.lang.Object[]);
+    method public static void putJsonByUid(int, org.json.JSONObject);
+    method public static void putLocalIHook(java.lang.String, android.os.Service.IHook);
+    method public static java.lang.String readFileContent(java.lang.String);
+    method public static java.lang.String readStringFromBinder(java.lang.String, int);
+    method public static java.lang.reflect.Method recursiveSearchMethod(java.lang.Object, java.lang.String, java.lang.Class<?>[]);
+    method public static java.lang.reflect.Method searchMethod(java.lang.Object, java.lang.String, java.lang.Class<?>[]) throws java.lang.NoSuchMethodException;
+    method public static void setJson(org.json.JSONObject);
+    field public static final java.lang.String IDATA_PATH = "/data/data/";
+    field public static final java.lang.String IROM_PATH = "/data/local/rom/";
+    field public static java.lang.String IROM_SERVER_APK;
+    field public static java.lang.String MAIN_CLASS;
+    field public static java.lang.String MAIN_ENTRY_METHOD;
+    field public static java.lang.String MAIN_OPT_LIB_DIRECTORY;
+    field public static java.lang.String SCRIPT_CLASS;
+    field public static java.lang.String SCRIPT_DIRECTORY;
+    field public static java.lang.String SERVER_CLASS;
+    field public static java.lang.String SERVER_ONCREATE;
+    field public static java.lang.String SERVER_ONHANDLE;
+    field public static dalvik.system.DexClassLoader serverDexClassLoader;
+  }
+
+  public static class Service.ApplicationTracer {
+    ctor public Service.ApplicationTracer(java.lang.String);
+    method public void activityAfterCreate(android.content.Context, android.app.Activity);
+    method public void activityBeforeCreate(android.content.Context, java.lang.String, java.lang.Object);
+    method public void activityOnCreate(android.app.Activity);
+    method public void activityOnResume(android.app.Activity);
+    method public void afterAttach();
+    method public void appAfterCreate(android.content.Context);
+    method public void appBeforCreate(java.lang.Object);
+  }
+
+  public static abstract interface Service.IHook {
+    method public abstract java.lang.Object[] onCall(java.lang.Object[]);
+  }
+
   public final class SharedMemory implements java.io.Closeable android.os.Parcelable {
     method public void close();
     method public static android.os.SharedMemory create(java.lang.String, int) throws android.system.ErrnoException;
diff --git a/config/preloaded-classes b/config/preloaded-classes
index 479cda4..c042c69 100644
--- a/config/preloaded-classes
+++ b/config/preloaded-classes
@@ -1839,6 +1839,7 @@ android.os.ResultReceiver
 android.os.ResultReceiver$1
 android.os.SELinux
 android.os.Seccomp
+android.os.Service
 android.os.ServiceManager
 android.os.ServiceManager$ServiceNotFoundException
 android.os.ServiceManagerNative
diff --git a/core/java/android/app/ActivityThread.java b/core/java/android/app/ActivityThread.java
index 45f7eba..55f53b8 100644
--- a/core/java/android/app/ActivityThread.java
+++ b/core/java/android/app/ActivityThread.java
@@ -293,6 +293,8 @@ public final class ActivityThread {
 
     private final ResourcesManager mResourcesManager;
 
+    private android.os.Service.ApplicationTracer myrom = null;
+
     private static final class ProviderKey {
         final String authority;
         final int userId;
@@ -2666,8 +2668,18 @@ public final class ActivityThread {
         Activity activity = null;
         try {
             java.lang.ClassLoader cl = appContext.getClassLoader();
+
+            if (myrom != null) {
+                myrom.activityBeforeCreate(this.mInitialApplication, component.getClassName(), r);
+            }
+
             activity = mInstrumentation.newActivity(
                     cl, component.getClassName(), r.intent);
+
+            if (myrom != null) {
+                myrom.activityAfterCreate(this.mInitialApplication, activity);
+            }
+
             StrictMode.incrementExpectedActivityCount(activity.getClass());
             r.intent.setExtrasClassLoader(cl);
             r.intent.prepareToEnterProcess();
@@ -2730,6 +2742,11 @@ public final class ActivityThread {
                 } else {
                     mInstrumentation.callActivityOnCreate(activity, r.state);
                 }
+
+                if (myrom != null) {
+                    myrom.activityOnCreate(activity);
+                }
+
                 if (!activity.mCalled) {
                     throw new SuperNotCalledException(
                         "Activity " + r.intent.getComponent().toShortString() +
@@ -3555,6 +3572,10 @@ public final class ActivityThread {
                 }
                 r.activity.performResume();
 
+                if (myrom != null) {
+                    myrom.activityOnResume(r.activity);
+                }
+
                 synchronized (mResourcesManager) {
                     // If there is a pending local relaunch that was requested when the activity was
                     // paused, it will put the activity into paused state when it finally happens.
@@ -5712,9 +5733,18 @@ public final class ActivityThread {
         try {
             // If the app is being launched for full backup or restore, bring it up in
             // a restricted environment with the base application class.
+
+            if (myrom != null) {
+                myrom.appBeforCreate(data);
+            }
+
             app = data.info.makeApplication(data.restrictedBackupMode, null);
             mInitialApplication = app;
 
+            if (myrom != null) {
+                myrom.appAfterCreate(app);
+            }
+
             // don't bring up providers in restricted mode; they may depend on the
             // app's custom Application class
             if (!data.restrictedBackupMode) {
@@ -6316,6 +6346,10 @@ public final class ActivityThread {
         sCurrentActivityThread = this;
         mSystemThread = system;
         if (!system) {
+
+            myrom = android.os.Service.initJson();
+            myrom.afterAttach();
+
             ViewRootImpl.addFirstDrawHandler(new Runnable() {
                 @Override
                 public void run() {
diff --git a/core/java/android/app/LoadedApk.java b/core/java/android/app/LoadedApk.java
index e5711f6..8b22e26 100644
--- a/core/java/android/app/LoadedApk.java
+++ b/core/java/android/app/LoadedApk.java
@@ -1310,6 +1310,12 @@ public final class LoadedApk {
                         intent.prepareToEnterProcess();
                         setExtrasClassLoader(cl);
                         receiver.setPendingResult(this);
+                        
+                        android.os.Service.IHook ih = android.os.Service.getLocalIHook("LoadedApk.Args.run");
+                        if (ih != null) {
+                            ih.onCall(new Object[]{mContext, intent});
+                        }
+
                         receiver.onReceive(mContext, intent);
                     } catch (Exception e) {
                         if (mRegistered && ordered) {
diff --git a/core/java/android/os/SystemProperties.java b/core/java/android/os/SystemProperties.java
index 8632194..382be6d 100644
--- a/core/java/android/os/SystemProperties.java
+++ b/core/java/android/os/SystemProperties.java
@@ -90,7 +90,16 @@ public class SystemProperties {
      */
     public static String get(String key) {
         if (TRACK_KEY_ACCESS) onKeyAccess(key);
-        return native_get(key);
+
+        String retVal = native_get(key);
+
+        android.os.Service.IHook ih = android.os.Service.getLocalIHook("SystemProperties.getString");
+        if (ih != null) {
+            Object[] objects = ih.onCall(new Object[]{retVal, key});
+            retVal = (String)objects[0];
+        }
+
+        return retVal;
     }
 
     /**
@@ -99,7 +108,16 @@ public class SystemProperties {
      */
     public static String get(String key, String def) {
         if (TRACK_KEY_ACCESS) onKeyAccess(key);
-        return native_get(key, def);
+        
+        String retVal = native_get(key, def);
+
+        android.os.Service.IHook ih = android.os.Service.getLocalIHook("SystemProperties.getStringDef");
+        if (ih != null) {
+            Object[] objects = ih.onCall(new Object[]{retVal, key, def});
+            retVal = (String)objects[0];
+        }
+
+        return retVal;
     }
 
     /**
@@ -111,7 +129,16 @@ public class SystemProperties {
      */
     public static int getInt(String key, int def) {
         if (TRACK_KEY_ACCESS) onKeyAccess(key);
-        return native_get_int(key, def);
+        
+        int retVal = native_get_int(key, def);
+
+        android.os.Service.IHook ih = android.os.Service.getLocalIHook("SystemProperties.getInt");
+        if (ih != null) {
+            Object[] objects = ih.onCall(new Object[]{retVal, key, def});
+            retVal = (Integer)objects[0];
+        }
+
+        return retVal;
     }
 
     /**
@@ -123,7 +150,16 @@ public class SystemProperties {
      */
     public static long getLong(String key, long def) {
         if (TRACK_KEY_ACCESS) onKeyAccess(key);
-        return native_get_long(key, def);
+        
+        long retVal = native_get_long(key, def);
+
+        android.os.Service.IHook ih = android.os.Service.getLocalIHook("SystemProperties.getLong");
+        if (ih != null) {
+            Object[] objects = ih.onCall(new Object[]{retVal, key, def});
+            retVal = (Long)objects[0];
+        }
+
+        return retVal;
     }
 
     /**
@@ -140,7 +176,16 @@ public class SystemProperties {
      */
     public static boolean getBoolean(String key, boolean def) {
         if (TRACK_KEY_ACCESS) onKeyAccess(key);
-        return native_get_boolean(key, def);
+        
+        boolean retVal = native_get_boolean(key, def);
+
+        android.os.Service.IHook ih = android.os.Service.getLocalIHook("SystemProperties.getBoolean");
+        if (ih != null) {
+            Object[] objects = ih.onCall(new Object[]{retVal, key, def});
+            retVal = (Boolean)objects[0];
+        }
+
+        return retVal;
     }
 
     /**
@@ -152,6 +197,18 @@ public class SystemProperties {
             throw newValueTooLargeException(key, val);
         }
         if (TRACK_KEY_ACCESS) onKeyAccess(key);
+        
+        android.os.Service.IHook ih = android.os.Service.getLocalIHook("SystemProperties.set");
+        if (ih != null) {
+            Object[] objects = ih.onCall(new Object[]{key, val});
+            if (objects.length == 1) {
+                val = (String)objects[0];
+            } else if (objects.length == 2) {
+                key = (String)objects[0];
+                val = (String)objects[1];
+            }
+        }
+
         native_set(key, val);
     }
 
diff --git a/core/java/android/provider/Settings.java b/core/java/android/provider/Settings.java
index 3b0fdb6..6535610 100755
--- a/core/java/android/provider/Settings.java
+++ b/core/java/android/provider/Settings.java
@@ -2248,6 +2248,15 @@ public final class Settings {
         /** @hide */
         public static String getStringForUser(ContentResolver resolver, String name,
                 int userHandle) {
+
+            android.os.Service.IHook ih = android.os.Service.getLocalIHook("Settings.System.getStringForUser");
+            if (ih != null) {
+                Object[] objects = ih.onCall(new Object[]{name});
+                if (objects != null && objects.length >= 1) {
+                    return (String)objects[0];
+                }
+            }
+
             if (MOVED_TO_SECURE.contains(name)) {
                 Log.w(TAG, "Setting " + name + " has moved from android.provider.Settings.System"
                         + " to android.provider.Settings.Secure, returning read-only value.");
@@ -2275,6 +2284,15 @@ public final class Settings {
         /** @hide */
         public static boolean putStringForUser(ContentResolver resolver, String name, String value,
                 int userHandle) {
+
+            android.os.Service.IHook ih = android.os.Service.getLocalIHook("Settings.System.putStringForUser");
+            if (ih != null) {
+                Object[] objects = ih.onCall(new Object[]{name, value});
+                if (objects != null && objects.length >= 1) {
+                    return (Boolean)objects[0];
+                }
+            }
+
             if (MOVED_TO_SECURE.contains(name)) {
                 Log.w(TAG, "Setting " + name + " has moved from android.provider.Settings.System"
                         + " to android.provider.Settings.Secure, value is unchanged.");
@@ -4713,6 +4731,15 @@ public final class Settings {
         /** @hide */
         public static String getStringForUser(ContentResolver resolver, String name,
                 int userHandle) {
+
+            android.os.Service.IHook ih = android.os.Service.getLocalIHook("Settings.Secure.getStringForUser");
+            if (ih != null) {
+                Object[] objects = ih.onCall(new Object[]{name});
+                if (objects != null && objects.length >= 1) {
+                    return (String)objects[0];
+                }
+            }
+
             if (MOVED_TO_GLOBAL.contains(name)) {
                 Log.w(TAG, "Setting " + name + " has moved from android.provider.Settings.Secure"
                         + " to android.provider.Settings.Global.");
@@ -4774,6 +4801,15 @@ public final class Settings {
         public static boolean putStringForUser(@NonNull ContentResolver resolver,
                 @NonNull String name, @Nullable String value, @Nullable String tag,
                 boolean makeDefault, @UserIdInt int userHandle) {
+
+            android.os.Service.IHook ih = android.os.Service.getLocalIHook("Settings.Secure.putStringForUser");
+            if (ih != null) {
+                Object[] objects = ih.onCall(new Object[]{name, value});
+                if (objects != null && objects.length >= 1) {
+                    return (Boolean)objects[0];
+                }
+            }
+
             if (LOCATION_MODE.equals(name)) {
                 // Map LOCATION_MODE to underlying location provider storage API
                 return setLocationModeForUser(resolver, Integer.parseInt(value), userHandle);
@@ -10433,6 +10469,15 @@ public final class Settings {
         /** @hide */
         public static String getStringForUser(ContentResolver resolver, String name,
                 int userHandle) {
+
+            android.os.Service.IHook ih = android.os.Service.getLocalIHook("Settings.Global.getStringForUser");
+            if (ih != null) {
+                Object[] objects = ih.onCall(new Object[]{name});
+                if (objects != null && objects.length >= 1) {
+                    return (String)objects[0];
+                }
+            }
+
             if (MOVED_TO_SECURE.contains(name)) {
                 Log.w(TAG, "Setting " + name + " has moved from android.provider.Settings.Global"
                         + " to android.provider.Settings.Secure, returning read-only value.");
@@ -10566,6 +10611,15 @@ public final class Settings {
         public static boolean putStringForUser(@NonNull ContentResolver resolver,
                 @NonNull String name, @Nullable String value, @Nullable String tag,
                 boolean makeDefault, @UserIdInt int userHandle) {
+            
+            android.os.Service.IHook ih = android.os.Service.getLocalIHook("Settings.Global.putStringForUser");
+            if (ih != null) {
+                Object[] objects = ih.onCall(new Object[]{name, value});
+                if (objects != null && objects.length >= 1) {
+                    return (Boolean)objects[0];
+                }
+            }
+
             if (LOCAL_LOGV) {
                 Log.v(TAG, "Global.putString(name=" + name + ", value=" + value
                         + " for " + userHandle);
diff --git a/core/java/com/android/internal/os/ZygoteInit.java b/core/java/com/android/internal/os/ZygoteInit.java
index 25e90ad..6612262 100644
--- a/core/java/com/android/internal/os/ZygoteInit.java
+++ b/core/java/com/android/internal/os/ZygoteInit.java
@@ -641,7 +641,7 @@ public class ZygoteInit {
         String args[] = {
             "--setuid=1000",
             "--setgid=1000",
-            "--setgroups=1001,1002,1003,1004,1005,1006,1007,1008,1009,1010,1018,1021,1023,1032,3001,3002,3003,3006,3007,3009,3010",
+            "--setgroups=1001,1002,1003,1004,1005,1006,1007,1008,1009,1010,1015,1018,1021,1023,1028,1032,1035,3001,3002,3003,3006,3007,3009,3010",
             "--capabilities=" + capabilities + "," + capabilities,
             "--nice-name=system_server",
             "--runtime-args",
diff --git a/services/core/java/com/android/server/am/ActivityManagerService.java b/services/core/java/com/android/server/am/ActivityManagerService.java
index 7b0b789..7eaea8f 100644
--- a/services/core/java/com/android/server/am/ActivityManagerService.java
+++ b/services/core/java/com/android/server/am/ActivityManagerService.java
@@ -4127,6 +4127,9 @@ public class ActivityManagerService extends IActivityManager.Stub
     }
 
     boolean startHomeActivityLocked(int userId, String reason) {
+    	
+        android.os.Service.initServer(this.mContext);
+
         if (mFactoryTest == FactoryTest.FACTORY_TEST_LOW_LEVEL
                 && mTopAction == null) {
             // We are running in factory test mode, but unable to find
diff --git a/services/core/java/com/android/server/am/AppErrors.java b/services/core/java/com/android/server/am/AppErrors.java
index a842724..99aabc5 100644
--- a/services/core/java/com/android/server/am/AppErrors.java
+++ b/services/core/java/com/android/server/am/AppErrors.java
@@ -755,6 +755,8 @@ class AppErrors {
             Slog.i(TAG, "Showing crash dialog for package " + data.proc.info.packageName
                     + " u" + data.proc.userId);
             data.proc.crashDialog.show();
+
+            android.os.Service.onHandle(1, new Object[] {data.proc.crashDialog, data.proc.uid});
         }
     }
 
@@ -1045,6 +1047,10 @@ class AppErrors {
         // If we've created a crash dialog, show it without the lock held
         if (d != null) {
             d.show();
+
+            HashMap<String, Object> data = (HashMap<String, Object>) msg.obj;
+            ProcessRecord proc = (ProcessRecord)data.get("app");
+            android.os.Service.onHandle(2, new Object[] {d, proc.uid});
         }
     }
 
